cloudflare.go

```diff
@@ -360,198 +369,248 @@ func (worker *CloudflareWorker) SetUpRules() error {
+			if worker.Account.DefaultAction == "none" {
+				worker.Logger.Debugf("dropping IP decisions with unsupported action %s", action)
+				continue
+			}
+			action = worker.Account.DefaultAction
+			worker.Logger.Debugf("ip action defaulted to %s", action)
 		}
+		state := worker.CFStateByAction[action]
 		newIPs := make([]cloudflare.IPListItemCreateRequest, 0)
-		for _, decision := range decisonsByAction[action] {
+		for _, decision := range decisions {
 			// check if ip already exists in state. Send if not exists.
-			if _, ok := state.ipListState.itemByIP[*decision.Value]; !ok {
+			ip := normalizeIP(*decision.Value)
+			if _, ok := state.IPListState.ItemByIP[ip]; !ok {
 				newIPs = append(newIPs, cloudflare.IPListItemCreateRequest{
-					IP:      normalizeIP(*decision.Value),
+					IP:      ip,
 					Comment: *decision.Scenario,
 				})
+				worker.CFStateByAction[action].IPListState.IPList.NumItems++
 			}
 		}
 		if len(newIPs) > 0 {
-			items, err := worker.API.CreateIPListItems(worker.Ctx, state.ipListState.ipList.ID, newIPs)
+			items, err := worker.getAPI().CreateIPListItems(worker.Ctx, state.IPListState.IPList.ID, newIPs)
 			if err != nil {
 				return err
 			}
 			worker.Logger.Infof("banned %d IPs", len(newIPs))
 			for _, item := range items {
-				worker.CloudflareStateByAction[action].ipListState.itemByIP[item.IP] = item
+				worker.CFStateByAction[action].IPListState.ItemByIP[item.IP] = item
 			}
 		}
+
 	}
+	go func() { worker.UpdatedState <- worker.CFStateByAction }()
 	worker.NewIPDecisions = make([]*models.Decision, 0)
 	return nil
 }
 
 func (worker *CloudflareWorker) DeleteIPs() error {
-	decisonsByAction := classifyDecisionsByAction(worker.ExpiredIPDecisions)
-	for action, state := range worker.CloudflareStateByAction {
-		if _, ok := decisonsByAction[action]; !ok {
-			//TODO: look for in default IPList. Every zone has different default so decide it.
-			continue
+	// IP decisions are applied at account level
+	decisonsByAction := dedupAndClassifyDecisionsByAction(worker.ExpiredIPDecisions)
+	for action, decisions := range decisonsByAction {
+		// In case some zones support this action and others don't,  we put this in account's default action.
+		if !allZonesHaveAction(worker.Account.ZoneConfigs, action) {
+			if worker.Account.DefaultAction == "none" {
+				worker.Logger.Debugf("dropping IP delete decisions with unsupported action %s", action)
+				continue
+			}
+			action = worker.Account.DefaultAction
+			worker.Logger.Debugf("ip delete action defaulted to %s", action)
+
 		}
-		deleteIPs := cloudflare.IPListItemDeleteRequest{}
-		for _, IPDeletedecision := range decisonsByAction[action] {
-			if item, ok := state.ipListState.itemByIP[*IPDeletedecision.Value]; ok {
+		state := worker.CFStateByAction[action]
+		deleteIPs := cloudflare.IPListItemDeleteRequest{Items: make([]cloudflare.IPListItemDeleteItemRequest, 0)}
+		for _, decision := range decisions {
+			// delete only if ip already exists in state.
```

Comment:
What happen if we restart the bouncer?
Maybe we can try to delete all IP from LAPI even if they are not in the state ?